### PR TITLE
Set the duration on the tvshow object

### PR DESCRIPTION
**Pull request by razzeee**
This gets a the runtime from the newest episode (that is not a special) and attaches it to the tvshow.
We will honor streamdetail runtime over scraped runtime.
This way skins can show a shows runtime via listitem.duration.

Adds another computed column to the tvshow_view, not really happy about that but best we can do right now.
- Needs mysql testing
- Performance consideration (shouldn't be that bad, at least not on sqlite)

### CODE REVIEW

**Comment by Razzeee** [comment]
If we want another approach, we could add a database field and just scrape the value into it.

**Comment by phate89** [comment]
thetvdb has a runtime field in their api. Why don't add a field to tvshow and scrape it?
Imho is more reliable. recently a lot of  tvshows have the first episode longer than the others...
You could actually place a fallback like:
`IFNULL(tvshow.duration, (SELECT COALESCE(s.iVideoDuration, e.c%02d) from episode e))`

**Comment by phil65** [comment]
For info, TheMovieDB also has some duration info available for TVShows (API can return an array containing one or more "common" durations of the TVShow) Not sure if we can process that with our scrapers though.

**Comment by Razzeee** [comment]
As I think tvdb only has that one runtime field, the scraped one on episodes is already of the same value. 
And if we know the file length of the newest episode that should even be better than those scraped values.

**Comment by BigNoid** [comment]
What about double episodes, eg Series.S01E01E02.mkv. Those have double duration in streamdetails.

**Comment by Razzeee** [comment]
Not sure how to get around that. I actually prefere knowing how long the file is as I will watch a file to the end and not stop after the first episode. 
Tvdb doesn't care either. For example http://thetvdb.com/?tab=series&id=75545&lid=7 this shouldn't be 30 min as those are mostly shorter.

**Comment by BigNoid** [comment]
Imo the tvshow duration should be the timeslot duration it has on the network, so a scraped value. If you have dvd boxset iso's, you need to name them as Series.S01E01E02E03E04.iso in order to get them as episodes listed. With this PR the TV show duration would be set to 2 hours for example.

**Comment by Razzeee** [comment]
@BigNoid 
your right. I'll change it to ignore the stream value then. But I will still fetch the scraped value from one of the episodes for now.

**Comment by Razzeee** [comment]
Updated to only get the scraped runtime.
I should probably ping @MilhouseVH for a mysql test?

**Comment by Tolriq** [comment]
I think scraped value (nfo / tvdb) + db field is better for long term than this solution :( 

It slows down tvshows query and the value is approximate since it can change for every episode depending on source, credits or not, intro or not, downloaded from the net or proper bd rip.

Scraped value also brings some consistency to the values, as a 1 hours show will be 40/45 minutes for US shows due to all those ads and 50/55 in other countries. 

This allow more easily to create smart playlist to filter shows by nominal duration.

But having this value is definitively something that would be cool to have :)

**Comment by Razzeee** [comment]
Well at this moment the implemented code is using the scraped value. And as tmdb and tvdb don't know per episode/season runtime, it will stay the same. As all episode entries right now have the same scraped runtime.

**Comment by Tolriq** [comment]
Or not : https://github.com/xbmc/xbmc/pull/5789 :)

And it still slower the basic get shows query.

**Comment by phate89** [comment]
IMHO it doesn't make sense to do extra work for every query when we can store the result and allow it to fix it via api/nfo..
If the episode use that value is the episode approach that is wrong..

**Comment by Razzeee** [comment]
I'll rework it then. My fear is that tvshow runtime will become invalid over time.

Like scrap when only season one is out. The runtime is 30min. But it's been such a success, that the network says, season two will be 45min long. I don't think our scrapers will update those values then.

**Comment by Razzeee** [comment]
So here are the changes, complete with json-rpc set method.

If you want to test it with the tvdb scraper, just add:
```			<RegExp input="$$5" output="&lt;runtime&gt;\1&lt;/runtime&gt;" dest="4+">
				<expression>&lt;Runtime&gt;([^&lt;]+)&lt;/Runtime&gt;</expression>
			</RegExp>```

in line 73 to your tvdb.xml 

@olympia 
Just so you know this is comming

**Comment by Tolriq** [comment]
@Razzeee just nitpicking and I have no strong feeling about that but I always find disturbing the different naming currently about duration and runtime depending on music / video.

Your PR expose runtime but internally use duration, is there a real need to use different terms ? Maybe use this PR to start uniforming the naming ? (Specially about external JSON API) is there a reason to use those different terms or is it just continuing what was done long ago ?

**Comment by Razzeee** [comment]
@tolriq 
Yeah, felt the same way. But I don't want to break the api. Will have to check if it would be okay to rename internally. 

@phate89 
Will do. But as tolriq pointed out of might lead to some stream times being in there. But shouldn't be a problem.

**Comment by Razzeee** [comment]
@Tolriq 
had a look and renaming that stuff is unfortunatly out of scope for this PR

@phate89 
we're not importing the max duration that's set on one of the episodes of a show, when migrating

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by Razzeee** [comment]
Jenkins builds we're fine on all platforms

**Comment by Razzeee** [comment]
Has anyone tested this so far? I would like to get it merged tomorrow if possible.
@phate89

File: xbmc/video/VideoDatabase.cpp [hunk 1]
```diff
@@ -4905,11 +4909,25 @@ void CVideoDatabase::UpdateTables(int iVersion)
     m_pDS->exec("ALTER TABLE settings ADD VideoStream integer");
     m_pDS->exec("ALTER TABLE streamdetails ADD strVideoLanguage text");
   }
+
+  if (iVersion < 104)
+  {
+    m_pDS->exec("ALTER TABLE tvshow ADD duration INTEGER");
+
+    std::string sql = PrepareSQL("SELECT idShow, MAX(c%02d) FROM episode GROUP BY idShow", VIDEODB_ID_EPISODE_RUNTIME);
```
Comment (phate89) [inline xbmc/video/VideoDatabase.cpp]: imho max isn't a good choice. There are a lot of first episodes of season/series longer than the regular episode.. I think it's better to use the latest episode's one..
I'll try to test the pr tomorrow

Comment (Razzeee) [inline xbmc/video/VideoDatabase.cpp]: I feel like showing a too long runtime is better than showing a too short runtime.
Like I have to go somewehere in 30mins, I would probably like to watch something  shorter than 30 min.
Just for the mindset.

**Comment by Razzeee** [comment]
Testbuilds started in order to see if MAX runtime when migrating brings out weird stuff, when there are streamdetails.
http://jenkins.kodi.tv/view/Helpers/job/BuildMulti-PR-Manually/208/

File: xbmc/video/VideoDatabase.cpp (hunk 1 above)
Comment (phate89) [inline xbmc/video/VideoDatabase.cpp]: Ok longer but this is way more. Every tvshow that have a 2 in 1 episode will have twice the duration, every episode that have a season/series finale longer will get a way longer runtime
Try this in your database:
`select tvshow.c00, MAX(episode.c09)/60 from tvshow JOIN episode ON episode.idShow = tvshow.idShow GROUP BY tvshow.idShow`

Comment (phate89) [inline xbmc/video/VideoDatabase.cpp]: I think we can separate the upgrade in 2 queries. The first is something like this:
`SELECT episode.idShow, MAX(episode.c09) from episode LEFT JOIN streamdetails ON episode.idFile = streamdetails.idFile WHERE episode.c09 <> streamdetails.iVideoDuration GROUP BY episode.idShow`
That gathers the runtime set from the scraper (if there is). The second is this query that fills the missing ones with avg/max/latest (whatever we want). In this way we get more reliable results

Comment (Razzeee) [inline xbmc/video/VideoDatabase.cpp]: I'm confused as we choose to stay away from streamdetails previously in this PR and taking this data now seems weird. I would like to stick to the scraped values, even if streamdetails are already bleeding into them. Note, that tvdb doesn't know of the runtime of each episode of a multiepisode, but only the complete block.
I still think this is the way to go and would really dislike us splitting these values, as I watch on a file base, not on an episode base.

**Comment by Razzeee** [comment]
@phate89 
updated, I settled for a best of both worlds approach. In my collection we would end up with only 51 of 130 tvshows having a runtime, if we only use the sql statement for scraped data. So I added a fallback to streamdata in case the scraped is missing.

File: xbmc/video/VideoDatabase.cpp [hunk 2]
```diff
@@ -4905,11 +4909,33 @@ void CVideoDatabase::UpdateTables(int iVersion)
...
+
+  if (iVersion < 104)
+  {
+    m_pDS->exec("ALTER TABLE tvshow ADD duration INTEGER");
+
+    std::string sql = PrepareSQL("SELECT e.idShow, COALESCE(MAX(e.c%02d),e2.c%02d) "
+                                 "FROM episode AS e "
+                                 "LEFT JOIN streamdetails AS s "
+                                 "ON e.idFile = s.idFile "
+                                 "AND e.c%02d <> s.iVideoDuration "
```
Comment (phate89) [inline xbmc/video/VideoDatabase.cpp]: You need to limit it only to video streams..

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by Razzeee** [comment]
Builds were fine, except for the usual windows addon problem.

**Pull request merged**